keps/sig-api-machinery/4933-cacher-label-index/README.md

```diff
@@ -0,0 +1,880 @@
+
+The matching and filtering in phase 2 and phase 3 are to determine whether a pod change is a change that a cacheWatcher is concerned about. The reason why it is divided into two phases is to optimize the performance of event processing as much as possible, with coarse-grained filtering in phase 2 and precise matching in phase 3. Since the current cacher only supports one trigger index for event processing of the incomming channel, only filtering according to a hard-coded field selector is supported, which is the hard-coded [`spec.nodeName`](https://github.com/kubernetes/kubernetes/blob/v1.31.2/pkg/registry/core/pod/storage/storage.go#L91) for pods. For the optimization effect in phase 2, please refer to [Implement first version of indexing in cacher](https://github.com/kubernetes/kubernetes/pull/27277)
+
+In a k8s cluster that mainly runs online applications, most of the watch requests are initiated by kubelet and daemonset programs, most of which watch via field selector `spec.nodeName`, so setting the pod's trigger index to `spec.nodeName` is efficient enough. But in a k8s cluster that mainly runs big data tasks, the situation is different.
+
+Each production k8s cluster in our company may run thousands of spark or flink jobs in parallel (the number of jobs may be roughly the same size as the number of nodes, or even higher). Each spark/flink job will start a master pod to create/list/watch/delete worker pods, which is Spark's `Driver` pod and Flink's `JobManager` pod, respectively. We found that this will cause a lot of pressure on the apiserver. In extreme cases, the number of pod add events sent by the apiserver to the scheduler per second will not keep up with the speed of pod creation per second, resulting in the scheduling speed not keeping up with the speed of pod creation, causing resource waste.
+
+Because a Spark or a Flink job only lists and watches the pod change events of itself by applicationId label, and the number of list and watch requests from these jobs is also comparable to the number of that from daemonsets in a considerable cluster. If we can have a label index as a cache index and trigger index, we can improve apiserver performance for running Spark and Flink jobs a lot.
+
+Although [PR#115918](https://github.com/kubernetes/kubernetes/pull/115918) has added namespace indexes to the apiserver cacher, this is not so useful in scenarios where a large number of jobs are running in a small number of namespaces. Most events are still sent to each cacheWatcher through the cacher first and then filtered, causing high CPU usage on the apiserver and delays in event delivery.
+
+### Background 1
+
+Each job of the big data computing engine spark/flink will start a master pod to create/list/watch/delete worker pods, namely the Driver pod of spark and the JobManager pod of flink.
```

Comment:
We avoid using “master” in this kind of context. In countries that have outlawed slavery, but with a history of allowing it, the term is sensitive even today.